### PR TITLE
Remove unnecessary parentheses from some if conditions

### PR DESCRIPTION
**Pull request by knocte**
First proof of concept for https://github.com/fsprojects/fantomas/issues/684

@nojaf let us know if this is a good beginning?

(PS: this might be a good change to postpone to after v4.4.0 final release, though)

### CODE REVIEW

**Comment by auduchinok** [comment]
@knocte JFYI if we're to remove more parens during formatting, it'll be very soon when may get into territory where removing parens might break code due to how nesting constructs are parsed in FCS due to different precedences, additional implicit blocks (due to indentation), blocks allowing deindent (due to parens) and so on. The last one seems to be a smaller issue due to Fantomas effectively creates new code with that has some expected indentation but we should take it into account anyway if we try to make more similar rules.

@nojaf It would be great to have such code modifications optional via settings due to:

* it should be enabled via some experimental flags if more sophisticated rules are added (it's relatively easy to make something wrong in parens analysis in F#)
* it's questionable if such cleanup/linter modification should be done during formatting stage

**Comment by knocte** [comment]
>  it'll be very soon when may get into territory where removing parens might break code due to how nesting constructs are parsed in FCS due to different precedences,

Well, my opinion is that if in some scenario X fantomas might break code by removing a parenthesis, then fantomas should not remove the parenthesis in scenario X. This is a PoC PR that removes parenthesis in only a very specific case that is impossible to break code: it just removes the outer parenthesis of an if condition (in F# it's not needed while in C# it is, so this is very low hanging fruit to fix and many codebases plagued with ex-C# devs mannerisms would be grateful of that -- NB: I include myself in that group of ex-C# devs that always forgets this).

> it should be enabled via some experimental flags 

That's fine by me, and also the reason why I was suggesting to not merge this before v4.4.0 final is released.

**Comment by knocte** [comment]
PS: I also think that there will be scenarios where it's impossible to break code by removing parenthesis but still would not be desirable for fantomas to remove them: e.g. transforming `x + (y * z)` to `x + y * z` would not bring a change of behaviour, however, the developer might have chosen to use the parenthesis to increase readability (e.g. towards F# newbies, or unexperienced programmers in general), so fantomas should never decrease readability.

**Comment by auduchinok** [comment]
> This is a PoC PR that removes parenthesis in only a very specific case that is impossible to break code: it just removes the outer parenthesis of an if condition

Some artificial examples can show that it's not always possible to remove these particular parens:
```fsharp
let a = true
let b = true
let foo = id

if (a) then ()

if (foo a) then ()

if (if true then a else b) then ()

if (match () with _ -> a) then ()

```

The last two examples don't look like common ways to write F# code but there're some examples like these out there in users code (especially the `match` case), and editor tools must never break it.

Another concern is there're a bit different rules for different places in the `if` expression, so this is allowed:

```fsharp
if true then match () with _ -> ()
```

> Well, my opinion is that if in some scenario X fantomas might break code by removing a parenthesis, then fantomas should not remove the parenthesis in scenario X.

Yes, I totally get it. I just wanted to stress that it may be difficult to analyze such scenarios unless a very strict allowed cases list is used or a comprehensive precedence/block analysis is done.

**Comment by knocte** [comment]
> if (foo a) then ()

This PR doesn't even remove parens in the case above, so I doubt it will remove them in the more complex ones you brought up.

**Comment by knocte** [comment]
It's weird that the CI status of the PR shows yellow as if it was pending. But FYI it is green if you go directly to the branch: https://github.com/knocte/fantomas/tree/removeParensFromIfConditions

PS: Yes we know we need to rebase it, will do it later.

File: src/Fantomas/CodePrinter.fs [hunk 1]
```diff
@@ -49,6 +51,7 @@ type ASTContext =
           IsFirstTypeParam = false
           IsInsideDotGet = false
           IsInsideDotIndexed = false
+          IsInsideIfThenElseExpression = false
```
Comment (nojaf) [inline src/Fantomas/CodePrinter.fs:54]: Adding new members to `ASTContext` is a recipe for disaster.
In recent weeks I removed a couple of entries, so I won't allow any new members here.

File: src/Fantomas/CodePrinter.fs [hunk 2]
```diff
@@ -1637,6 +1640,17 @@ and genExpr astContext synExpr ctx =
         | Paren (_, ILEmbedded r, _) ->
             // Just write out original code inside (# ... #)
             fun ctx -> !- (defaultArg (lookup r ctx) "") ctx
+
+        | Paren (lpr, SynExpr.Ident ident, rpr) when astContext.IsInsideIfThenElseExpression ->
+            let e = SynExpr.Ident ident
+
+            match ident.idText with
+            | "op_Multiply" ->
```
Comment (nojaf) [inline src/Fantomas/CodePrinter.fs:1648]: This is not really the safest check. See [example](https://fsprojects.github.io/fantomas-tools/#/ast?data=N4KABGBEAmCmBmBLAdrAzpAXFSAacUiaAYmolmPAIYA2as%2BEkAxgPZwWQ2wAuYrABwD6AWQCuNHogE0AnmAC8YACwAmSCAC%2BQA).

File: src/Fantomas/CodePrinter.fs [hunk 3]
```diff
@@ -1637,6 +1640,17 @@ and genExpr astContext synExpr ctx =
         | Paren (_, ILEmbedded r, _) ->
             // Just write out original code inside (# ... #)
             fun ctx -> !- (defaultArg (lookup r ctx) "") ctx
+
+        | Paren (lpr, SynExpr.Ident ident, rpr) when astContext.IsInsideIfThenElseExpression ->
```
Comment (nojaf) [inline src/Fantomas/CodePrinter.fs:1644]: I prefer this check to happen inside the `Elif` match clause.
`genExpr` is called all over the place.

**Review by nojaf — CHANGES_REQUESTED** [review]
Hello Andres, thank you for taking the time to investigate this feature request.
I'm very anxious towards this concept to be honest.
It is introducing additional complexity for a tool that still has [a lot of issues to solve](https://github.com/fsprojects/fantomas/issues?q=is%3Aissue+is%3Aopen+label%3A%22bug+%28soundness%29%22).

The sample you provided is a nice illustration of how complex things can possibly get.
What other operators will be need to detect? What about [trivia that is assigned to parenthesis](https://fsprojects.github.io/fantomas-tools/#/trivia?data=N4KABGBEDGD2AmBTSAuKAbRAXMBDMAvGADoBOxAdhBAJYBmYAFGAPQtgBissZl1-EAEa5SYAJRgsAC0QVeVAdQC2iKZAA04KEjo0KiAM6ooGrZBoGOBmsbq50BxCAC%2BQA)?
At some point the information in the Syntax Tree might also not suffice to make the correct call whether the parenthesis can be removed or not.

I'm also not 100% convinced that is the responsibility of a code formatter to have this. Having it in Fantomas is just "a solution" to tackle the problem you have. Did you consider adding a refactor feature to the editor you use?

I'll give this some more thought, but at the time of writing I don't feel like pursuing this functionality.

**Comment by knocte** [comment]
> It is introducing additional complexity for a tool that still has a lot of issues to solve.

Would wrapping this behaviour in an opt-in setting (as @auduchinok suggested) give you more peace of mind? This way at least further bugfixes to fantomas may not be harder to fix because they would not need to be aware of the code flows that happen when this feature is enabled. My team would be willing to maintain this feature.

> What about trivia that is assigned to parenthesis? ...so I won't allow any new members here.

We'll have a look at your points and we'll update the PR to address them (adding more tests too).

Thanks

**Comment by nojaf** [comment]
Hello @knocte, I still haven't heard any solid arguments why this should be part of Fantomas and not some editor tooling?
You still need a stronger story here.

I'd have more peace of mind if this was not at all in Fantomas rather than behind a setting.

**Comment by lydell** [comment]
> some editor tooling

What editor tooling would that be, as an example? (I’m not sure what “editor tooling” means.)

I think this repo would benefit from such tooling. Check out these unnecessary parens:

```
src/Fantomas/TokenParserBoolExpr.fs
47:            | BoolExpr.Not (BoolExpr.Not (e)) -> Some e

src/Fantomas/SourceTransformer.fs
139:    | MDMember (x1) :: MDMember (x2) :: xs ->

src/Fantomas/Queue.fs
28:                Seq.forall2 (Unchecked.equals) this y

src/Fantomas/TriviaHelpers.fs
74:                | true, Some (t) -> predicate t
82:                | true, Some (StringContent (s)) when (String.isMultiline s) -> Some s

src/Fantomas/CodeFormatterImpl.fs
32:    | SourceString s -> FSharp.Compiler.Text.SourceText.ofString (s)
119:        | SynModuleDecl.NamespaceFragment (fragment) -> validateModuleOrNamespace fragment
268:        | SynExpr.Ident (_ident) -> true
302:        | SynExpr.Null (_range)
303:        | SynExpr.ImplicitZero (_range) -> true
459:    | Some (tok), state ->
471:    | Some (tok), state ->
688:/// Format a part of source string using given config, and return the (formatted) selected part only.

src/Fantomas/SourceParser.fs
282:    | String (s) when (String.isMultiline s) -> Some e
1658:            |> Option.map (Char.IsUpper)
1667:    | SynExpr.Ident (Ident (id)) -> upperOrLower id
1676:    | SynExpr.DotGet (_, _, LongIdentWithDots (lid), _) -> upperOrLower lid
1687:            |> Option.map (Char.IsUpper)
1717:    | App (OptVar (ident), [ ArrayOrList _ as attributes; ArrayOrList (isArray, children, _) as childrenNode ]) ->
1719:    | App (OptVar (ident),
1723:    | App (OptVar (ident),
1726:    | App (OptVar (ident),
1729:    | App (OptVar (ident), [ ArrayOrListOfSeqExpr _ as attributes; ArrayOrList (isArray, [], _) as childrenNode ]) ->

src/Fantomas/TriviaContext.fs
86:                | Some (IdentBetweenTicks (iiw)) -> Some iiw
87:                | Some (IdentOperatorAsWord (iow)) -> Some iow

src/Fantomas/AstTransformer.fs
155:        | SynModuleDecl.NamespaceFragment (moduleOrNamespace) ->
426:        | SynExpr.Ident (id) ->
547:        | SynExpr.Null (range) ->
578:        | SynExpr.ImplicitZero (range) ->
844:        | SynTypeDefnSigRepr.Exception (exceptionRepr) ->
1128:        | SynPat.Wild (range) ->
1212:        | SynPat.Null (range) ->
1263:        | Pats (pats) ->
1289:                    yield! (typeParams |> List.map (visitSynTyparDecl)) ] }
1307:        | SynTypeDefnRepr.Exception (exceptionRepr) ->
1387:        | SynTypeDefnSimpleRepr.None (range) ->
1433:        | SynTypeDefnSimpleRepr.Exception (edr) ->
1518:        | UnionCaseFields (cases) ->
1567:        | SynType.LongIdent (li) ->
1627:        | SynType.Anon (range) ->
1858:        | SynModuleSigDecl.NamespaceFragment (moduleOrNamespace) ->

src/Fantomas/Context.fs
215:                    | MainNode (mn) -> Some(mn, tn)
363:    | Write (w) when (String.startsWithOrdinal "//" w) -> Some we
364:    | Write (w) when (String.startsWithOrdinal "#if" w) -> Some we
365:    | Write (w) when (String.startsWithOrdinal "#else" w) -> Some we
366:    | Write (w) when (String.startsWithOrdinal "#endif" w) -> Some we
367:    | Write (w) when (String.startsWithOrdinal "(*" w) -> Some we
442:/// `atCurrentColumn` was called on `X`, then `indent` was called, but "some long string" have indent only 4, because it is bigger than `atColumn` (2) [... 0 more matches]
456:/// `atCurrentColumn` was called on `X`, then `indent` was called, "some long string" have indent 6, because it is indented from `atCurrentColumn` pos [... 1 more match]
527:        st <- f i (e.Current) st
546:        st <- f (e.Current) st
562:        st <- f (e.Current) st
782:    (fallbackExpression)
823:let internal isShortExpression maxWidth (shortExpression: Context -> Context) (fallbackExpression) (ctx: Context) =
1138:    | Directive (s)
1208:        | Some ({ ContentAfter = [ TriviaContent.Comment (LineCommentAfterSourceCode (lineComment)) ] }) ->
1226:        | Some ({ ContentBefore = [ TriviaContent.Comment (LineCommentOnSingleLine (lineComment)) ] }) ->
1365:    | Some range when (ctx.Config.NewlineBetweenTypeDefinitionAndMembers) ->
1429:                | Some (fr1) when (fr1 = r1) ->
1443:            | Some (fr1) when (fr1 = r) ->
1488:                        | TriviaContent.Comment (Comment.LineCommentAfterSourceCode (comment)) -> Some comment

src/Fantomas/TokenParser.fs
196:                    |> fun chars -> new string (chars)
512:    | StringTextToken (stToken) :: InterpStringEndOrPartToken (endToken) :: rest -> Some([ stToken ], endToken, rest)
513:    | StringTextToken (stToken) :: EndOfInterpolatedString (stringTokens, endToken, rest) ->
807:    | CharToken (head) :: rest ->
872:                | StringContent (sc) when (isMultilineString sc) -> Some tc.Range

src/Fantomas/Trivia.fs
12:    | MainNode (t) when (t <> SynModuleOrNamespace_AnonModule) -> true
17:    | MainNode (SynModuleOrNamespace_AnonModule)
18:    | MainNode (SynModuleOrNamespaceSig_AnonModule) -> true
28:    | MainNode (t) when (t = nodeType) -> true
68:    | MainNode (mn) -> mn = name
133:            | MainNode (SynMemberDefn_Member), true
134:            | MainNode (SynMemberSig_Member), true
157:                    | MainNode (mn) when (mn = SynExpr_Ident) ->
186:            | MainNode (SynExpr_Const)
187:            | MainNode (SynPat_Const) ->
190:            | MainNode (EnumCase_) ->
200:            | MainNode (SynConst_String), true, true -> true
234:                | MainNode (SynModuleOrNamespace_NamedModule)
235:                | MainNode (SynModuleOrNamespaceSig_NamedModule) -> true
262:            | MainNode (SynPat_Named) when
265:            | MainNode (SynPat_LongIdent) when
268:            | MainNode (SynExpr_Ident) when
278:            | MainNode (ParsedHashDirective_) when
307:            | TriviaNodeType.MainNode (mnt) when (mnt = typeName) -> RangeHelpers.``range contains`` t.Range range
403:                | MainNode (SynMemberSig_Member), Some (Keyword ({ Content = existingKeywordContent } as token)) when
446:    | { Item = Keyword (keyword)
458:    | { Item = Directive (dc) as directive
503:                    | MainNode (SynExpr_Ident)
504:                    | MainNode (SynPat_Named)
505:                    | MainNode (SynPat_LongIdent)
506:                    | MainNode (Ident_)
507:                    | MainNode (SynConst_String) -> true

src/Fantomas.CoreGlobalTool/Program.fs
72:    |> Set.exists (fullPath.Contains)
90:            && not (IgnoreFile.isIgnoredFile (f)))
210:    | InputPath.File (path) ->
215:    | InputPath.Folder (path) ->

src/Fantomas/CodePrinter.fs
371:    | Attributes (ats) ->
402:            (attributesExpr) ctx
403:    | DoExpr (e) -> genExpr astContext e
404:    | Exception (ex) -> genException astContext ex
405:    | HashDirective (p) -> genParsedHashDirective p
426:    | Let (b) -> genLetBinding { astContext with IsFirstChild = true } "let " b
451:    | NamespaceFragment (m) -> failwithf "NamespaceFragment hasn't been implemented yet: %O" m
465:    | Open (s) -> !-(sprintf "open %s" s)
466:    | OpenType (s) -> !-(sprintf "open type %s" s)
489:    | SigException (ex) -> genSigException astContext ex
490:    | SigHashDirective (p) -> genParsedHashDirective p
491:    | SigVal (v) -> genVal astContext v
493:    | SigNamespaceFragment (m) -> failwithf "NamespaceFragment is not supported yet: %O" m
505:    | SigOpen (s) -> !-(sprintf "open %s" s)
506:    | SigOpenType (s) -> !-(sprintf "open type %s" s)
1431:        | CompExprBody (expr) ->
1766:                    | TypeApp (LongIdentPieces (lids), ts) when (List.moreThanOne lids) ->
1794:                | LongIdentPieces (lids) when (List.moreThanOne lids) -> genFunctionNameWithMultilineLids f lids
1795:                | TypeApp (LongIdentPieces (lids), ts) when (List.moreThanOne lids) ->
1822:                | AppOrTypeApp (LongIdentPieces (lids), ts, [ e2 ]) when (List.moreThanOne lids) ->
1863:                    +> ifElseCtx (addSpace) sepSpace sepNone
1869:                    +> ifElseCtx (addSpace) sepSpace sepNone
1983:                        | Sequentials (s) -> s |> List.collect synExpr
2250:                        | AppSingleArg _ when (isConditionMultiline) ->
2254:                        | Paren (_, (AppSingleArg _ as ate), _) when (isConditionMultiline) ->
2510:                                    | MainNode (SynInterpolatedStringPart_String), Some (StringContent sc) when
3369:    | ObjectModel (TCSimple (TCStruct) as tdk, MemberDefnList (impCtor, others), _) ->
3775:                            |> optSingle (fun (Ident (s)) -> !-s +> sepColon)
3902:        | TStaticConstantExpr (e) -> genExpr astContext e
3973:        | TParen (innerT) ->
4207:    | MDOpen (s) -> !-(sprintf "open %s" s)
4251:    | MDMember (b) -> genMemberBinding astContext b
4315:                        | Some (Keyword (kw)) -> Some kw.Content
4413:    | PatOptionalVal (s) -> !-(sprintf "?%s" s)
4425:    | PatAnds (ps) -> col (!- " & ") ps (genPat astContext)
4493:    | PatParen (p) ->
4523:    | PatRecord (xs) ->
4559:    | PatIsInst (t) ->
4940:    | SynConst.Bool (b) -> !-(if b then "true" else "false")
4968:                        | Some (StringContent (sc)) -> Some sc
4984:    | SynConst.Char (c) ->
5018:                | Some (Number (n)) -> Some n
5025:                | SynConst.Byte (v) -> !-(sprintf "%A" v)
5026:                | SynConst.SByte (v) -> !-(sprintf "%A" v)
5027:                | SynConst.Int16 (v) -> !-(sprintf "%A" v)
5028:                | SynConst.Int32 (v) -> !-(sprintf "%A" v)
5029:                | SynConst.Int64 (v) -> !-(sprintf "%A" v)
5030:                | SynConst.UInt16 (v) -> !-(sprintf "%A" v)
5031:                | SynConst.UInt16s (v) -> !-(sprintf "%A" v)
5032:                | SynConst.UInt32 (v) -> !-(sprintf "%A" v)
5033:                | SynConst.UInt64 (v) -> !-(sprintf "%A" v)
5034:                | SynConst.Double (v) -> !-(sprintf "%A" v)
5035:                | SynConst.Single (v) -> !-(sprintf "%A" v)
5036:                | SynConst.Decimal (v) -> !-(sprintf "%A" v)
5037:                | SynConst.IntPtr (v) -> !-(sprintf "%A" v)
5038:                | SynConst.UIntPtr (v) -> !-(sprintf "%A" v)
5051:                    | Some (StringContent (content)) -> Some content
5082:                | Some (IdentOperatorAsWord (iiw)) -> Some iiw
5083:                | Some (IdentBetweenTicks (iiw)) when not (isValidIdent fallback) -> Some iiw // Used when value between ``...``
```

**Comment by knocte** [comment]
> What editor tooling would that be, as an example?

I was about to raise same point, cause the closest I see where this could be solved is FSharpLint, however I see this tool more "semantic" (e.g. it barks at you if you call a partial function), and IMO this "unnecessary parentheses" feature is more of a syntax issue.

**Comment by nojaf** [comment]
I don't want to pursue this in this project.
Please consider contributing to:
- JetBrains Rider (https://github.com/JetBrains/fsharp-support)
- Ionide /VsCode (https://github.com/fsharp/FsAutoComplete)
- Visual Studio (https://github.com/dotnet/fsharp)

**Pull request closed without merging**